ColorField docs page and more stories (#1631)
adobe/react-spectrum: packages/@react-spectrum/color/docs/ColorField.mdx

of the License at http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License. -->

import {Layout} from '@react-spectrum/docs';
export default Layout;

import docs from 'docs:@react-spectrum/color';
import {HeaderInfo, PropTable} from '@react-spectrum/docs';
import packageData from '@react-spectrum/color/package.json';

```jsx import
import {ColorField} from '@react-spectrum/color';
import {Flex} from '@react-spectrum/layout';
```

---
category: Color
keywords: [color field, input]
---

# ColorField

<p>{docs.exports.ColorField.description}</p>

<HeaderInfo
  packageData={packageData}
  componentNames={['ColorField']}
  sourceData={[]} />

## Example

```tsx example
<ColorField label="Primary Color" />
```

## Value

A ColorField's `value` is empty by default, but an initial, uncontrolled, value can be provided using the `defaultValue` prop.

> Alternatively, a controlled value can be provided using the `value` prop. These values must begin with a `#`.

Reviewer: Do we wanna mention that it should be a valid hex color code? (maybe a bit too obvious tho)

Author: Good question, I wasn't sure if mentioning that they have to start with a # is specific enough or not... saying a valid hex color code would probably be more specific and better.

```tsx example
function Example() {
  let [value, setValue] = React.useState('#e73623');

  return (
    <Flex gap="size-150" wrap>
      <ColorField
        label="Primary Color (Uncontrolled)"
        defaultValue="#e73623" />

      <ColorField
        label="Primary Color (Controlled)"
        value={value}
        onChange={setValue} />
    </Flex>
  );
}
```

> Placeholder text, provided using the placeholder prop, describes the expected value or formatting for the ColorField.

Reviewer: ```suggestion
Placeholder text, provided using the `placeholder` prop, describes the expected value or formatting for the ColorField.
```

component with a visible label.

<ColorField placeholder="#e73623" label="Primary Color" />

## Labeling

A visual label should be provided for the ColorField using the `label` prop. If the ColorField is required, the `isRequired` and
`necessityIndicator` props can be used to show a required state.

```tsx example
<Flex gap="size-150" wrap>
  <ColorField label="Primary Color" />
  <ColorField label="Primary Color" isRequired />
  <ColorField label="Primary Color" isRequired necessityIndicator="label" />
  <ColorField label="Primary Color" necessityIndicator="label" />
</Flex>
```

### Accessibility

If a visible label isn't specified, an `aria-label` must be provided to the ColorField for
accessibility. If the field is labeled by a separate element, an `aria-labelledby` prop must be provided using
the `id` of the labeling element instead.

### Internationalization

In order to internationalize a ColorField, a localized string should be passed to the `label` or `aria-label` prop.
When the `necessityIndicator` prop is set to `"label"`, a localized string will be provided for `"(required)"` or `"(optional)"` automatically.

> ## Step values

Author: This behavior is different than NumberInput step and was documented how it works. Confirming that we want this difference.

> The `step` prop can be used to snap the value to certain increments, default is one. The steps are calculated starting from
> zero or the current value. A `step` can be any positive decimal.

Reviewer: Flow nit
```suggestion
The `step` prop can be used to snap the value to certain increments. The steps are calculated starting from
zero or the current value. A `step` can be set to any positive decimal, defaulting to `1` if unspecified .
```

```tsx example

> <NumberField label="Primary Color" step={16} />

Reviewer: ```suggestion
<ColorField label="Primary Color" step={16} />
```

```

## Props

> <PropTable component={docs.exports.ColorField} links={docs.links} />

Reviewer: Missing a description for `isQuiet`

> 

Reviewer: What do you think about adding a `validation` section here to bring it in line with TextField/NumberField/etc docs?

## Visual options

### Quiet

```tsx example
<ColorField label="Primary Color" isQuiet />

> ```

Reviewer: The examples aren't showing up in the built version of the docs, not sure why. Might be the NumberField -> ColorField example above that is affecting these?

accessibility. If the field is labeled by a separate element, an `aria-labelledby` prop must be provided using
the `id` of the labeling element instead.

### Internationalization

In order to internationalize a ColorField, a localized string should be passed to the `label` or `aria-label` prop.
When the `necessityIndicator` prop is set to `"label"`, a localized string will be provided for `"(required)"` or `"(optional)"` automatically.

## Step values

The `step` prop can be used to snap the value to certain increments. The steps are calculated starting from
zero or the current value. A `step` can be set to any positive decimal, defaulting to `1` if unspecified.

```tsx example
<ColorField label="Primary Color" step={16} />
```

## Validation

ColorField can display a validation state to communicate to the user whether the current value is valid or invalid.
Implement your own validation logic in your app and pass either `"valid"` or `"invalid"` to the ColorField via the `validationState` prop.

The example below illustrates how one would validate if the user has entered a red color.
```tsx example
function Example() {

>   let [value, setValue] = React.useState('#e73623');

Author: If I use the stately color package like Daniel did https://github.com/adobe/react-spectrum/pull/1632/files#diff-addf8a33459e32a12f6d20f3f399163936469198e52834f7413162debace95d2R57, I could get rid of the if statement, simplifying the example. I choose to not do that because I wasn't sure if people reading this example and creating an app using React Spectrum would do that too. The context of what Daniel is showing is slightly different, but it still makes me unsure what is the best choice here.

Reviewer 2: Yes we want people to take advantage of the parse function. I would add a popover link in the value section or something? Something like this https://reactspectrum.blob.core.windows.net/reactspectrum/d5112eb2e9b59888ec5217353bf1fd83566a6c03/docs/react-aria/useColorWheel.html#controlled

## Example

```tsx example
<ColorField label="Primary Color" />
```

## Value

A ColorField's `value` is empty by default, but an initial, uncontrolled, value can be provided using the `defaultValue` prop.
Alternatively, a controlled value can be provided using the `value` prop. Valid values are hexadecimal color codes.
The <TypeLink links={statelyDocs.links} type={statelyDocs.exports.parseColor} /> function is used to parse the initial color
from a hex value, stored in state.

```tsx example
import {parseColor} from '@react-stately/color';

function Example() {
  let [value, setValue] = React.useState(parseColor('#e73623'));

  return (
    <Flex gap="size-150" wrap>
      <ColorField
        label="Primary Color (Uncontrolled)"
        defaultValue="#e21" />

      <ColorField
        label="Primary Color (Controlled)"
        value={value}
        onChange={setValue} />
    </Flex>
  );
}
```

Placeholder text, provided using the `placeholder` prop, describes the expected value or formatting for the ColorField.
Placeholder text will only appear when the ColorField is empty, and should not be used as a substitute for labeling the
component with a visible label.

> <ColorField placeholder="#e73623" label="Primary Color" />

Author: I'll remove this section, but it's listed as a prop and it does work. Do you want me to create an issue to remove it?

## Labeling

A visual label should be provided for the ColorField using the `label` prop. If the ColorField is required, the `isRequired` and
`necessityIndicator` props can be used to show a required state.

```tsx example
<Flex gap="size-150" wrap>
  <ColorField label="Primary Color" />
  <ColorField label="Primary Color" isRequired />
  <ColorField label="Primary Color" isRequired necessityIndicator="label" />
  <ColorField label="Primary Color" necessityIndicator="label" />
</Flex>
```

### Accessibility

If a visible label isn't specified, an `aria-label` must be provided to the ColorField for
accessibility. If the field is labeled by a separate element, an `aria-labelledby` prop must be provided using
the `id` of the labeling element instead.

### Internationalization

In order to internationalize a ColorField, a localized string should be passed to the `label` or `aria-label` prop.
When the `necessityIndicator` prop is set to `"label"`, a localized string will be provided for `"(required)"` or `"(optional)"` automatically.

## Step values

The `step` prop can be used to snap the value to certain increments. The steps are calculated starting from

> zero or the current value. A `step` can be set to any positive decimal, defaulting to `1` if unspecified.

Author: In numberfield it snaps to a step interval, have a min of 2, step 5, a value of 10 changes to 12.
In colorfield the value doesn't change based on the step. A step of 32, the value never changes or snaps to an interval of that step.

## Step values

The `step` prop can be used to snap the value to certain increments. The steps are calculated starting from
zero or the current value. A `step` can be set to any positive decimal, defaulting to `1` if unspecified.

```tsx example
<ColorField label="Primary Color" step={16} />
```

## Validation

> NumberFields can display a validation state to communicate to the user whether the current value is valid or invalid.
> Implement your own validation logic in your app and pass either `"valid"` or `"invalid"` to the NumberField via the `validationState` prop.

Reviewer: ```suggestion
ColorFields can display a validation state to communicate to the user whether the current value is valid or invalid.
Implement your own validation logic in your app and pass either `"valid"` or `"invalid"` to the ColorField via the `validationState` prop.
```

The example below illustrates how one would validate if the user has entered a red color.
```tsx example
function Example() {
  let [value, setValue] = React.useState('#e73623');
  let isValid = React.useMemo(() => {

>     if (value.length === 7) {

Reviewer: Noticed that the example breaks if the user deletes everything in the field, think you have to make sure value is defined here too.
